NitroxPatcher/Patches/Dynamic/Constructable_Construct_Patch.cs

```diff
@@ -1,62 +1,185 @@
+using System.Linq;
+using System.Reflection;
+using HarmonyLib;
+using NitroxClient.Communication.Abstract;
+using NitroxClient.GameLogic.Bases;
+using NitroxClient.GameLogic.Spawning.Bases;
+using NitroxClient.Helpers;
+using NitroxClient.MonoBehaviours;
+using NitroxModel.DataStructures;
+using NitroxModel.DataStructures.GameLogic;
+using NitroxModel.DataStructures.GameLogic.Bases;
+using NitroxModel.DataStructures.GameLogic.Entities;
+using NitroxModel.DataStructures.GameLogic.Entities.Bases;
 using NitroxModel.Helper;
+using NitroxModel.Packets;
+using NitroxPatcher.PatternMatching;
+using UnityEngine;
+using UWE;
+using static System.Reflection.Emit.OpCodes;
+using static NitroxClient.GameLogic.Bases.BuildingHandler;
 
-namespace NitroxPatcher.Patches.Dynamic
+namespace NitroxPatcher.Patches.Dynamic;
+
+public sealed partial class Constructable_Construct_Patch : NitroxPatch, IDynamicPatch
 {
-    public sealed partial class Constructable_Construct_Patch : NitroxPatch, IDynamicPatch
+    public static readonly MethodInfo TARGET_METHOD = Reflect.Method((Constructable t) => t.Construct());
+
+    private static TemporaryBuildData Temp => BuildingHandler.Main.Temp;
+
+    public static readonly InstructionsPattern InstructionsPattern = new()
     {
-        public static readonly MethodInfo TARGET_METHOD = Reflect.Method((Constructable t) => t.Construct());
+        Div,
+        Stfld,
+        Ldc_I4_0,
+        Ret,
+        Ldarg_0,
+        { InstructionPattern.Call(nameof(Constructable), nameof(Constructable.UpdateMaterial)), "Insert" }
+    };
 
-        private static Base lastTargetBase;
-        private static Int3 lastTargetBaseOffset;
-        private static Base.Face lastFace;
+    public static readonly List<CodeInstruction> InstructionsToAdd = new()
+    {
+        new(Ldarg_0),
+        new(Call, Reflect.Method(() => ConstructionAmountModified(default)))
+    };
 
-        public static bool Prefix(Constructable __instance)
+    public static IEnumerable<CodeInstruction> Transpiler(MethodBase original, IEnumerable<CodeInstruction> instructions) =>
+        instructions.Transform(InstructionsPattern, (label, instruction) =>
         {
-            if (__instance.constructed)
+            if (label.Equals("Insert"))
             {
-                return true;
+                return InstructionsToAdd;
             }
+            return null;
+        });
 
-            // If we are constructing a base piece then we'll want to store all of the BaseGhost information
-            // as it will not be available when the construction hits 100%
-            if (__instance is ConstructableBase constructableBase)
+    public static void ConstructionAmountModified(Constructable constructable)
+    {
+        // We only manage the amount change, not the deconstruction/construction action
+        if (!constructable.TryGetNitroxId(out NitroxId entityId))
+        {
+            // TODO: Maybe delete the ghost
```

Comment:
Would write investigate or similar, not maybe. Maybe implies you are not confident of your code.

Reply:
For whatever reason this would happen, I wanted to manage the error but I'll just let a log in there.
I originally wrote this TODO when planing what resyncing features I'd add, and this ended up not being part of it.